packages/workflows/src/handlers/middlewares/update-products-extract-created-variants.ts

```diff
@@ -0,0 +1,42 @@
+import { ProductTypes, ProductVariantDTO } from "@medusajs/types"
+
+import { WorkflowArguments } from "../../helper"
+import { UpdateProductsPreparedData } from "../product"
+
+export async function updateProductsExtractCreatedVariants({
+  data,
+}: WorkflowArguments<{
+  preparedData: UpdateProductsPreparedData // products state before the update
+  products: ProductTypes.ProductDTO[] // updated products
+}>) {
+  const createdVariants: ProductVariantDTO[] = []
+
+  data.products.forEach((product) => {
+    const addedVariants: ProductVariantDTO[] = []
+
+    const originalProduct = data.preparedData.originalProducts.find(
+      (p) => p.id === product.id
+    )!
+
+    product.variants.forEach((variant) => {
+      if (!originalProduct.variants.find((v) => v.id === variant.id)) {
+        addedVariants.push(variant)
+      }
+    })
+
+    createdVariants.push(...addedVariants)
+  })
```

Comment:
todo: this should do the same and be constant in perf
```ts
import { ProductTypes } from "@medusajs/types"

import { WorkflowArguments } from "../../helper"
import { UpdateProductsPreparedData } from "../product"

export async function updateProductsExtractCreatedVariants({
  data,
}: WorkflowArguments<{
  preparedData: UpdateProductsPreparedData // products state before the update
  products: ProductTypes.ProductDTO[] // updated products
}>) {
  const originalVariantIds = new Set(
    data.preparedData.originalProducts.reduce((acc, product) => {
      const variantIds = (product.variants ?? []).map((v) => v.id)
      acc.push(...variantIds)
      return acc
    }, [] as string[])
  )

  const productVariantsMap = new Map<string, ProductTypes.ProductVariantDTO>()
  const productsVariantIds = new Set(
    data.products.reduce((acc, product) => {
      const variantIds = (product.variants ?? []).map((v) => v.id)
      acc.push(...variantIds)
      return acc
    }, [] as string[])
  )

  const createdVariantIds = Array.from(productsVariantIds).filter(
    (variant) => {
      return !originalVariantIds.has(variant)
    }
  )

  const createdVariants = createdVariantIds.map((variantId) => {
    return productVariantsMap.get(variantId)
  })

  return {
    alias: updateProductsExtractCreatedVariants.aliases.output,
    value: {
      variants: createdVariants,
    },
  }
}

updateProductsExtractCreatedVariants.aliases = {
  preparedData: "preparedData",
  products: "products",
  output: "updateProductsExtractCreatedVariantsOutput",
}

```

Reply:
In general, when working with multiple collection it is always good to think about the appropriate data structure to use in order to perform the action we want :) I haven't checked all the algo everywhere but I just want to remind that point. In my opinion it seems also easier to read and the maintain but I might be wrong. wdyt?

Reply:
I find it easy to read and we can see that each step is properly separated. But I think that is a matter of opinion and I can be wrong in that sense too ahah. I let you decide on that one 👍